# Libraries and loading the dataset
In this first part of the examples we provide we will look at how to load in the dataset, and how to do some basic
manipulation of the data. In order to run the code in this notebook you will need to install, besides Jupyter, the
[Pandas](https://pandas.pydata.org/docs/getting_started/index.html#installation) package.

For more info on how to set up the Python environment for these notebooks, see README.md.

## Getting an overview of the data
Whenever you get a dataset you will want to inspect it and see what variables (columns) you have and what these values
look like. One way of doing this is to use the package [Pandas](https://pandas.pydata.org/docs/getting_started/overview.html),
a package that allows for easy manipulation of the data table (or dataframe, as it is called in Pandas).

For a quick overview of Pandas' capabilities, check out [this](https://pandas.pydata.org/docs/user_guide/10min.html#min)
link.

In [50]:
import pandas as pd

data = pd.read_csv(r"../../data/owid-covid-data.csv")

# Checking table columns and data points
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 83862 entries, 0 to 83861
Data columns (total 59 columns):
 #   Column                                 Non-Null Count  Dtype  
---  ------                                 --------------  -----  
 0   iso_code                               83862 non-null  object 
 1   continent                              79805 non-null  object 
 2   location                               83862 non-null  object 
 3   date                                   83862 non-null  object 
 4   total_cases                            81878 non-null  float64
 5   new_cases                              81876 non-null  float64
 6   new_cases_smoothed                     80875 non-null  float64
 7   total_deaths                           72262 non-null  float64
 8   new_deaths                             72420 non-null  float64
 9   new_deaths_smoothed                    80875 non-null  float64
 10  total_cases_per_million                81435 non-null  float64
 11  ne

In the above output table you can see the indices of the columns (using standard 0-indexing) as well as the human-readable
name and information about what the data in those columns is. For example: at index 0 we find the `iso_index` this is a
shorthand code for countries (see [here](https://www.iso.org/obp/ui/#search/code/)).

We can also see that there are 83862 'non-nullable' rows for that column. This means non-empty cells in that column. Since
dataframes can be sparsely populated, some rows can have empty column values (e.g. a row can miss the value for `new_tests`
but still have other values).

Lastly, we can see the Python data type for that column. Most will either be floating point numbers or unnamed objects -
meaning they contain multiple values within a single data point.

## Fixing the date column
We can see that `date` column is not of the `datetime` object type, a commonly used format for timestamps in Python. It
will be useful to use this column as the x-axis for various plots and analyses, but most of those techniques assume that
the timestamps are in the aforementioned `datetime` format.

This means that if we want to use the `date` column effectively, we will have to alter the data type to `datetime`. The
following code block will show you how to do that. It is recommended to do this for your own analysis too!

In [51]:
data['date'] = pd.to_datetime(data['date'], format = '%Y-%m-%d')

Your dataset is now ready for your analysis! You may want to sort your data set by date in order to more intuitively see
changes. You can do so doing:

In [ ]:
# Note: `date` is the name of the column being sorted on.
data = data.sort_values('date', ascending=True)

## Taking a look at the dataframe
The next notebook will deal more with the topic of data visualisation and exploring the data set, but for now you can
visualize (part of) your dataframe by using the following commands:

In [52]:
# Show first N (e.g. 10) rows of the dataframe
data.head(10)

iso_code      continent   location       date  total_cases  new_cases  \
3009       ARG  South America  Argentina 2020-01-01          NaN        NaN   
48722      MEX  North America     Mexico 2020-01-01          NaN        NaN   
48723      MEX  North America     Mexico 2020-01-02          NaN        NaN   
3010       ARG  South America  Argentina 2020-01-02          NaN        NaN   
3011       ARG  South America  Argentina 2020-01-03          NaN        NaN   
48724      MEX  North America     Mexico 2020-01-03          NaN        NaN   
3012       ARG  South America  Argentina 2020-01-04          NaN        NaN   
75171      THA           Asia   Thailand 2020-01-04          NaN        NaN   
48725      MEX  North America     Mexico 2020-01-04          NaN        NaN   
75172      THA           Asia   Thailand 2020-01-05          NaN        NaN   

       new_cases_smoothed  total_deaths  new_deaths  new_deaths_smoothed  \
3009                  NaN           NaN         NaN                  NaN   
48722                 NaN           NaN         NaN                  NaN   
48723                 NaN           NaN         NaN                  NaN   
3010                  NaN           NaN         NaN                  NaN   
3011                  NaN           NaN         NaN                  NaN   
48724                 NaN           NaN         NaN                  NaN   
3012                  NaN           NaN         NaN                  NaN   
75171                 NaN           NaN         NaN                  NaN   
48725                 NaN           NaN         NaN                  NaN   
75172                 NaN           NaN         NaN                  NaN   

       total_cases_per_million  new_cases_per_million  \
3009                       NaN                    NaN   
48722                      NaN                    NaN   
48723                      NaN                    NaN   
3010                       NaN                    NaN   
3011                       NaN                    NaN   
48724                      NaN                    NaN   
3012                       NaN                    NaN   
75171                      NaN                    NaN   
48725                      NaN                    NaN   
75172                      NaN                    NaN   

       new_cases_smoothed_per_million  total_deaths_per_million  \
3009                              NaN                       NaN   
48722                             NaN                       NaN   
48723                             NaN                       NaN   
3010                              NaN                       NaN   
3011                              NaN                       NaN   
48724                             NaN                       NaN   
3012                              NaN                       NaN   
75171                             NaN                       NaN   
48725                             NaN                       NaN   
75172                             NaN                       NaN   

       new_deaths_per_million  new_deaths_smoothed_per_million  \
3009                      NaN                              NaN   
48722                     NaN                              NaN   
48723                     NaN                              NaN   
3010                      NaN                              NaN   
3011                      NaN                              NaN   
48724                     NaN                              NaN   
3012                      NaN                              NaN   
75171                     NaN                              NaN   
48725                     NaN                              NaN   
75172                     NaN                              NaN   

       reproduction_rate  icu_patients  icu_patients_per_million  \
3009                 NaN           NaN                       NaN   
48722                NaN           NaN                       NaN   
48

In [53]:
# Tail will show the last N results of the dataframe.
data.tail(10)

iso_code      continent  location       date  total_cases  new_cases  \
46479       MYS           Asia  Malaysia 2021-04-24     390252.0     2717.0   
41871       LAO           Asia      Laos 2021-04-24        247.0       88.0   
59539       PRY  South America  Paraguay 2021-04-24     265296.0     2162.0   
3935        ABW  North America     Aruba 2021-04-24          NaN        NaN   
57045  OWID_OCE            NaN   Oceania 2021-04-24      42985.0      276.0   
75170       TZA         Africa  Tanzania 2021-04-24        509.0        0.0   
28257       GAB         Africa     Gabon 2021-04-24      22433.0        0.0   
80366       URY  South America   Uruguay 2021-04-24     182326.0     2789.0   
21298       DJI         Africa  Djibouti 2021-04-24      10746.0        8.0   
83861       ZWE         Africa  Zimbabwe 2021-04-24      38064.0       19.0   

       new_cases_smoothed  total_deaths  new_deaths  new_deaths_smoothed  \
46479            2484.714        1426.0        11.0                8.000   
41871              27.000           NaN         NaN                0.000   
59539            2418.857        5802.0        87.0               78.714   
3935                  NaN           NaN         NaN                  NaN   
57045             203.571        1041.0         3.0                3.000   
75170               0.000          21.0         0.0                0.000   
28257              82.143         138.0         0.0                0.714   
80366            2846.571        2283.0        56.0               62.143   
21298              47.714         132.0         0.0                2.571   
83861              52.143        1556.0         0.0                0.571   

       total_cases_per_million  new_cases_per_million  \
46479                12057.468                 83.946   
41871                   33.949                 12.095   
59539                37195.217                303.118   
3935                       NaN                    NaN   
57045                 1007.198                  6.467   
75170                    8.521                  0.000   
28257                10078.949                  0.000   
80366                52487.141                802.884   
21298                10876.496                  8.097   
83861                 2561.003                  1.278   

       new_cases_smoothed_per_million  total_deaths_per_million  \
46479                          76.769                    44.059   
41871                           3.711                       NaN   
59539                         339.130                   813.456   
3935                              NaN                       NaN   
57045                           4.770                    24.392   
75170                           0.000                     0.352   
28257                          36.906                    62.002   
80366                         819.457                   657.219   
21298                          48.294                   133.603   
83861                           3.508                   104.690   

       new_deaths_per_million  new_deaths_smoothed_per_million  \
46479                   0.340                            0.247   
41871                     NaN                            0.000   
59539                  12.198                           11.036   
3935                      NaN                              NaN   
57045                   0.070                            0.070   
75170                   0.000                            0.000   
28257                   0.000                            0.321   
80366                  16.121                           17.889   
21298                   0.000                            2.603   
83861                   0.000                            0.038   

       reproduction_rate  icu_patients  icu_patients_per_million  \
46479                NaN           NaN                       NaN   
41871                NaN           NaN                       NaN   
595

In [54]:
# Array slicing also works - this shows rows 40 up to (not including) 45.
data[40:45]



iso_code      continent   location       date  total_cases  new_cases  \
48734      MEX  North America     Mexico 2020-01-13          NaN        NaN   
59547      PER  South America       Peru 2020-01-13          NaN        NaN   
75180      THA           Asia   Thailand 2020-01-13          NaN        NaN   
3021       ARG  South America  Argentina 2020-01-13          NaN        NaN   
59548      PER  South America       Peru 2020-01-14          NaN        NaN   

       new_cases_smoothed  total_deaths  new_deaths  new_deaths_smoothed  \
48734                 NaN           NaN         NaN                  NaN   
59547                 NaN           NaN         NaN                  NaN   
75180                 NaN           NaN         NaN                  NaN   
3021                  NaN           NaN         NaN                  NaN   
59548                 NaN           NaN         NaN                  NaN   

       total_cases_per_million  new_cases_per_million  \
48734                      NaN                    NaN   
59547                      NaN                    NaN   
75180                      NaN                    NaN   
3021                       NaN                    NaN   
59548                      NaN                    NaN   

       new_cases_smoothed_per_million  total_deaths_per_million  \
48734                             NaN                       NaN   
59547                             NaN                       NaN   
75180                             NaN                       NaN   
3021                              NaN                       NaN   
59548                             NaN                       NaN   

       new_deaths_per_million  new_deaths_smoothed_per_million  \
48734                     NaN                              NaN   
59547                     NaN                              NaN   
75180                     NaN                              NaN   
3021                      NaN                              NaN   
59548                     NaN                              NaN   

       reproduction_rate  icu_patients  icu_patients_per_million  \
48734                NaN           NaN                       NaN   
59547                NaN           NaN                       NaN   
75180                NaN           NaN                       NaN   
3021                 NaN           NaN                       NaN   
59548                NaN           NaN                       NaN   

       hosp_patients  hosp_patients_per_million  weekly_icu_admissions  \
48734            NaN                        NaN                    NaN   
59547            NaN                        NaN                    NaN   
75180            NaN                        NaN                    NaN   
3021             NaN                        NaN                    NaN   
59548            NaN                        NaN                    NaN   

       weekly_icu_admissions_per_million  weekly_hosp_admissions  \
48734                                NaN                     NaN   
59547                                NaN                     NaN   
75180                                NaN                     NaN   
3021                                 NaN                     NaN   
59548                                NaN                     NaN   

       weekly_hosp_admissions_per_million  new_tests  total_tests  \
48734                                 NaN      227.0       1598.0   
59547                                 NaN        1.0          3.0   
75180                                 NaN       14.0         64.0   
3021                                  NaN        2.0        295.0   
59548                                 NaN        NaN          NaN   

       total_tests_per_thousand  new_tests_per_thousand  new_tests_smoothed  \
48734                     0.012                   0.002               158.0   
59547                     0.000                   0.000                 0.0   
75180           